runtime/reconcilers/alert.go

```diff
@@ -0,0 +1,847 @@
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// refsStateHash computes a hash of the refs and their state versions.
+func (r *AlertReconciler) refsStateHash(ctx context.Context, refs []*runtimev1.ResourceName) (string, error) {
+	hash := md5.New()
+
+	for _, ref := range refs { // Refs are always sorted
+		// Write name
+		_, err := hash.Write([]byte(ref.Kind))
+		if err != nil {
+			return "", err
+		}
+		_, err = hash.Write([]byte(ref.Name))
+		if err != nil {
+			return "", err
+		}
+
+		// Note: Only writing the state version to the hash, not spec version, because it doesn't matter whether the spec/meta changes, only whether the state changes.
+		r, err := r.C.Get(ctx, ref, false)
+		var stateVersion int64
+		if err == nil {
+			stateVersion = r.Meta.StateVersion
+		} else {
+			stateVersion = -1
+		}
+		err = binary.Write(hash, binary.BigEndian, stateVersion)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// updateNextRunOn evaluates the alert's schedule relative to the current time, and updates the NextRunOn state accordingly.
+// If the schedule is nil, it will set NextRunOn to nil.
+func (r *AlertReconciler) updateNextRunOn(ctx context.Context, self *runtimev1.Resource, a *runtimev1.Alert) error {
+	next, err := nextRefreshTime(time.Now(), a.Spec.RefreshSchedule)
+	if err != nil {
+		return err
+	}
+
+	var curr time.Time
+	if a.State.NextRunOn != nil {
+		curr = a.State.NextRunOn.AsTime()
+	}
+
+	if next == curr {
+		return nil
+	}
+
+	if next.IsZero() {
+		a.State.NextRunOn = nil
+	} else {
+		a.State.NextRunOn = timestamppb.New(next)
+	}
+
+	return r.C.UpdateState(ctx, self.Meta.Name, self)
+}
+
+// setTriggerFalse sets the alert's spec.Trigger to false.
+// Unlike the State, the Spec may be edited concurrently with a Reconcile call, so we need to read and edit it under a lock.
+func (r *AlertReconciler) setTriggerFalse(ctx context.Context, n *runtimev1.ResourceName) error {
+	r.C.Lock(ctx)
+	defer r.C.Unlock(ctx)
+
+	self, err := r.C.Get(ctx, n, false)
+	if err != nil {
+		return err
+	}
+
+	a := self.GetAlert()
+	if a == nil {
+		return fmt.Errorf("not an alert")
+	}
+
+	a.Spec.Trigger = false
+	return r.C.UpdateSpec(ctx, self.Meta.Name, self)
+}
+
+// executeAll runs queries and (maybe) sends emails for the alert. It also adds entries to a.State.ExecutionHistory.
+// By default, an alert is checked once for the current watermark, but if a.Spec.IntervalsIsoDuration is set, it will be checked *for each* interval that has elapsed since the previous execution watermark.
+func (r *AlertReconciler) executeAll(ctx context.Context, self *runtimev1.Resource, a *runtimev1.Alert, triggerTime time.Time, adhocTrigger bool) error {
+	// Enforce timeout
+	timeout := alertCheckDefaultTimeout
+	if a.Spec.TimeoutSeconds > 0 {
+		timeout = time.Duration(a.Spec.TimeoutSeconds) * time.Second
```

Comment:
Should the timeout be a multiple of number of each block of interval this needs to run?

Reply:
Looking at the execution in depth, it will recover from last executed interval. But there will be a failure in reconcile no? Would that lead to resource being deleted and all the history being lost?

Reply:
Good questions. I made the timeout apply to all queries together since the purpose of it is to prevent runaway query execution that makes reconciles take a very long time.

About failure, yes it will recover and lead to a reconcile error, but that won't cause the resource to be deleted. The resource will still be there (and keep its state), it will just have `.Meta.ReconcileError` set until it is retried.